design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+
+To connect to Strimzi, a service binding needs the followng:
+
+* **host** - hostname or IP address
+* **port** - port number
+* **userName** - username, optional
+* **password** - password or token, optional
+* **certificate** - certificate, optional
+
+These pieces of information can be provided in a variety of ways:
+
+* annotations on custom resources, indicating `spec` and `status` properties
+* ConfigMap
+* Secret
+
+## Strimzi enhancement - Concatenation of bootstrap server information
+
+The main difficulty with using the Service Binding Operator with Strimzi is that there's no convenient way to get a consolidated list of bootstrap servers for the listener that you wish to use.
+
+The way to find the bootstrap information from the `Kafka` CR is to look at the status for the listeners. Here's an example:
+
+``` yaml
+apiVersion:kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 2.4.0
+    replicas: 2
+    listeners:
+      plain: {}
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 1
+    storage:
+      type: ephemeral
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+```
+
+To bind nicely to this, ideally we'd have a consolidated list `myhost1.example.com:9092,myhost2.example.com:9092` that can directly be used as the Kafka client's bootstrap server list. Most examples dodge this issue by just using the first address in the list, but that's a terrible idea for availability.
+
+One way to do this would be to add a `bootstrap` property to `status.listeners` like this:
+
+``` yaml
+apiVersion:kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+ ...
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+    bootstrap: myhost1.example.com:9092,myhost2.example.com:9092
+```
+
+Then the binding can use `status.listeners.plain.bootstrap`. If the listener name could be made into a parameter from the `ServiceBindingRequest`, then it would be possible to annotate the CSV for the `Kafka` CRD so that the service binding request would work without needing the user to annotate their custom resource.
+
+``` yaml
+apiVersion: operators.coreos/com:v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: strimzi-cluster-operator.v0.16.2
+  namespace: placeholder
+  annotations:
+    containerImage: docker.io/strimzi/operator:0.16.2
+spec:
+  customresourcedefinitions:
+    owned:
+    - description: Represents a Kafka cluster
+      displayName: Kafka
+      kind: Kafka
+      name: kafkas.kafka.strimzi.io
+      version: v1beta1
+      statusDescriptors:
+      - description: The addresses of the bootstrap servers for Kafka clients
+        displayName: Bootstrap servers
+        path: status.listeners.{{servicebinding:listener}}.bootstrap
```

Comment:
I'm not sure I understand why this is any relation to the Strimzi CSV. I do not think that should be in any way dependent on the ServiceBinding operator, or?

Reply:
I don't think this field makes Strimzi dependent on the ServiceBinding operator, it just provides some metadata that effectively flags it as usable by the ServiceBinding operator and also acts as the pointer for where to get credentials if/when the ServiceBinding operator gets a ServiceBinding request.

Reply:
The Service binding operator is optional -> so how would this resolve when it is and when it isn't installed? I think this is a dependency.

Reply:
My understanding is that it wouldn't be resolved before the csv is applied, it would be resolved by the Service Binding operator when it tries to do a binding. Although I think the format might need tweaking to make it clear that it isn't just a substitution error. So the csv would always show it exactly as above but the ServiceBinding operator would be able to use that info, combined with the info in a ServiceBindingRequest object to find the bootstrap address for the listener the app wants to bind to

Reply:
But as Andrew stated below "CSV has a limitation here in its descriptor path capability, so **OLM would also need enhancing**. Let's try another way."

The approach here matches most closely what the ServiceBinding operator have been expecting operators to do, but since Strimzi cannot point to a single location for a bootstrap address, that is why Andrew has proposed the the KafkaUser approach as an alternative.

Reply:
> My understanding is that it wouldn't be resolved before the csv is applied, it would be resolved by the Service Binding operator when it tries to do a binding.

This section of the CSV defines the UI which is visible to the user. So to add things there it needs to be clear how it renders or doesn't render in different situations.

Reply:
I personally would prefer not to need a variable substitution for listener name, and that's why I included a way of putting the listener name in the KafkaUser. I was just throwing this idea into the pot to see what people thought. I think people didn't like it :)

Reply:
Document now heavily revised. The proposal is essentially - a few small changes to the Strimzi Kafka CR to make it easier to obtain bootstrap server information, and annotations in the CSV which make Strimzi bindable by the Service Binding Operator.